src/PolyhedralGeometry/PolyhedralFan/properties.jl

```diff
@@ -303,10 +303,38 @@ julia> primitive_generator(r)
 ```
 """
 function primitive_generator(r::AbstractVector{T}) where {T<:RationalUnion}
-  result = numerator.(lcm([denominator(i) for i in r]) * r)
+  return Vector{ZZRingElem}(first(primitive_generator_with_scaling_factor(r)))
+end
+
+@doc raw"""
+    primitive_generator_with_scaling_factor(r::AbstractVector{T}) where T<:RationalUnion -> Tuple{Vector{ZZRingElem}, QQFieldElem}
+
+Given a vector `r` which is not a zero vector, returns the primitive generator `s` of `r` together with the rational number `q` such that `qr = s`.
+
+# Examples
+```jldoctest
+julia> PF = polyhedral_fan(IncidenceMatrix([[1, 2]]), [[2, -1], [0, 1]])
+Polyhedral fan in ambient dimension 2
+
+julia> r = rays(PF)[1]
+2-element RayVector{QQFieldElem}:
+ 1
+ -1//2
+
+julia> primitive_generator_with_scaling_factor(r)
+(ZZRingElem[2, -1], 2)
+```
+"""
+function primitive_generator_with_scaling_factor(
+  r::AbstractVector{T}
+) where {T<:RationalUnion}
+  first_scaling_factor = ZZ(lcm([denominator(i) for i in r]))
+  result = ZZ.(first_scaling_factor * r)
   g = gcd(result)
+  scaling_factor = QQ(first_scaling_factor)//QQ(g)
+  @assert g > 0 "The vector `r` cannot be a zero vector"
```

Comment:
```suggestion
  @req g > 0 "The vector `r` cannot be a zero vector"
  scaling_factor = QQ(first_scaling_factor, g)
```
Using `@req` is preferred over `@assert`, also the check should be before the division by zero.